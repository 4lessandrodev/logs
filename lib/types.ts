export type Type = "error" | "info" | "warn" | "debug" | "fatal" | "stack";
export type Locale = Intl.LocalesArgument;
export type LocalOpt = Intl.DateTimeFormatOptions | undefined;
export type Method = 'GET' | 'POST' | 'DELETE' | 'PATCH' | 'PUT' | 'OPTIONS' | 'HEAD' | 'LINK' | 'PURGE' | 'UNLINK' | 'NONE';

export interface SProps {
    readonly uid: string;
    readonly name: string;
    readonly tags: Readonly<string[]>;
    readonly url: string;
    readonly stack: string;
    readonly data: string;
    readonly statusCode: number;
    readonly message: string;
    readonly method: Method;
    readonly type: Type;
    readonly createdAt: Date;
}

export interface LProps {
    readonly uid: string;
    readonly name: string;
    readonly ip: string;
    readonly origin: string;
    readonly createdAt: Date;
    readonly steps: Readonly<Steps[]>;
}

export interface Steps extends SProps {
    addTag(tag: string): Readonly<Steps>;
    addTags(tags: string[]): Readonly<Steps>;
    setName(name: string): Readonly<Steps>;
    setMethod(method: Method): Readonly<Steps>;
    setStack(stack: string): Readonly<Steps>;
    setMessage(message: string): Readonly<Steps>;
    setStatusCode(code: number): Readonly<Steps>;
    setData(data: string): Readonly<Steps>;
    setUid(uid: string): Readonly<Steps>;
    setURL(url: string): Readonly<Steps>;
    print(locales?: Locale, options?: LocalOpt): void;
    getPrintableMsg(locales?: Locale, options?: LocalOpt): string;
}

export interface Logs extends LProps {
    setName(name: string): Readonly<Logs>;
    setIp(ip: string): Readonly<Logs>;
    setOrigin(url: string): Readonly<Logs>;
    addStep(step: Readonly<Steps>): Readonly<Logs>;
    addSteps(steps: Readonly<Steps[]>): Readonly<Logs>;
    removeStep(uid: string): Readonly<Logs>;
    writeLocal(path?: string): Promise<void>;
    print(locales?: Locale, options?: LocalOpt): void;
    publish(): Promise<void>;
}

export type BuildStepMessages = (step: Steps, locales?: Locale, options?: LocalOpt) => string;
export type BuildLogMessages = (log: Logs, locales?: Locale, options?: LocalOpt) => string;
type keys = Type | 'default';
export type FnTypes = { [k in keys ]: BuildStepMessages };
