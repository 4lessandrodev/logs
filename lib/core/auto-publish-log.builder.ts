import CanAutoPublish from "../utils/can-auto-publish.util";
import { AutoPublishOptions, NextFunctions, PublishConfig } from "../types";
import { PublisherMiddleware, Requests, Responses } from "../types";

/**
 * @description Add middleware to listen response and publish log when status match if options.
 * @param config connection configuration to connect on provider.
 * @param options option to check status before publish. You can define publish log only when status match with options.
 * @returns middleware to apply.
 * @default publishWhenStatus every statusCode.
 * @default publish only when exist some step on log.
 */
export const autoPublishLog = (config: PublishConfig, options?: AutoPublishOptions): PublisherMiddleware => {
    return async (req: Requests, res: Responses, next: NextFunctions): Promise<void> => {
        res.on('finish', async (): Promise<void> => { 
            const canPublish = CanAutoPublish(req, res, options);
            if(!canPublish) return next();
            await req?.log?.publish(config);
            return next();
        });
        return next();
    }
}
export default autoPublishLog;
