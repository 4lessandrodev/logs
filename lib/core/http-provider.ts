import * as axios from 'axios';
import { Http, HttpConfig, Logs, Provider, SavePayload } from "../types";

const HttProvider: Provider<HttpConfig> = ({
    async save(config: HttpConfig, log: Readonly<Logs>): Promise<SavePayload> {
        const http = axios as unknown as Http;
        const url = typeof config.url === 'string' ? new URL(config.url) : config.url;
        const result = await http.post(url.toString(), log, { headers: config.headers });
        return { statusCode: result.status ?? 200, url: url.toString() };
    }
});

export default HttProvider;
