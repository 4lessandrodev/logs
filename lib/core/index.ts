export * from './log';
export * from './step';
export * from './stack-log.builder';
export * from './bind-log.builder';
export * from './s3-provider';
export * from './http-provider';
export * from './providers';
export * from './auto-publish-log.builder';
export * from './global-log';
