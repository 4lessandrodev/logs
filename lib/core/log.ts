import { Color } from "../utils/color.util";
import { randomUUID } from "node:crypto";
import { Logs, LProps, Steps } from "../types";
import WriteDefaultLocal from "../utils/write-default-local.util";

export class Log implements Logs {
    readonly uid!: string;
    readonly name!: string;
    readonly ip!: string;
    readonly origin!: string;
    readonly createdAt!: Date;
    readonly steps!: Readonly<Steps[]>;

    private constructor(props: Partial<LProps>) {
        this.uid = props.uid ?? randomUUID();
        this.name = props.name ?? 'default';
        this.ip = props.ip ?? 'none';
        this.origin = props.origin ?? 'none';
        this.createdAt = new Date();
        this.steps = props.steps ?? [];
        Object.freeze(this);
    }

    public static init(props: Partial<LProps> & { name: string; uid: string }): Readonly<Logs> {
        return new Log(props);
    }

    /**
     * @description Create a new instance of Log with ip attribute. This is an immutable instance, the method does not change state, it returns a new one.
     * @param ip as request origin address
     * @returns instance of Log with ip address set
     */
    setIp(ip: string): Readonly<Logs> {
        return new Log({ ...this, ip });
    }

    /**
     * @description Create a new instance of Log with url attribute. This is an immutable instance, the method does not change state, it returns a new one.
     * @param url as request origin url address
     * @returns instance of Log with url address set
     */
    setOrigin(url: string): Readonly<Logs> {
        return new Log({ ...this, origin: url });
    }

    /**
     * @description Define log name. This attribute is used to create a folder name on store local.
     * @param name log name as string
     * @returns instance of Log with log name set
     */
    setName(name: string): Readonly<Logs> {
        return new Log({ ...this, name });
    }

    /**
     * @description Add a log step to instance.
     * @param step as instance of Step.
     * @returns instance of Log with added step.
     */
    addStep(step: Readonly<Steps>): Readonly<Logs> {
        return new Log({ ...this, steps: [...this.steps, step] });
    }

    /**
     * @description Add a log step to instance.
     * @param step as instance of Step.
     * @returns instance of Log with added step.
     */
    addSteps(steps: Readonly<Steps[]>): Readonly<Logs> {
        return new Log({ ...this, steps: [...this.steps, ...steps] });
    }

    /**
     * @description Remove a log step from instance.
     * @param uid as step uid to identify what step to remove.
     * @returns instance of Log without removed step.
     */
    removeStep(uid: string): Readonly<Logs> {
        const steps = this.steps.filter((step) => step.uid !== uid);
        return new Log({ ...this, steps });
    }

    /**
     * @description Save logs locally where the app is running
     * @param path absolute path of the folder where the file will be saved.
     * @returns 
     * @emits fileName: Omit the file name as the file name is defined based on the log name plus creation date
     * @example path: "/home/user/my-app/logs"
     * @default path: "/[app-folder]/logs/[log-name]-[year]-[month]-[day].txt".
     * @throws If an error occurs in the local saving process, the logs will be displayed on the standard output (terminal)
     */
    async writeLocal(path?: string): Promise<void> {
        try {
            return WriteDefaultLocal(this, path);
        } catch (error) {
            return this.print();
        }
    }

    print(): void {
        const { name, createdAt, steps, origin } = this;
        const time = createdAt.toLocaleString('pt');
        const message = ` Date: ${time} - ${name} - Origin: ${origin} `;
        const mainMsg = Color.white(message, 'magenta');
        const boldMain = Color.style().bold(mainMsg);
        const title = Color.style().reset(boldMain);
        const msgs = steps.map((step) => step.getPrintableMsg());
        const subMsg = msgs.map((msg) => `${msg}\n`).toString().replace(/,/g, '');
        console.log(`${title}\n${subMsg}`);
    }

    /**
     * @todo implement provider to publish on
     * @external firebase
     * @external aws-s3
     * @external mongodb
     * @external redis
     */
    async publish(): Promise<void> {
        console.log("publishing...");
    }
}

export default Log;
