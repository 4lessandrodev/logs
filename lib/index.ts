export * from './types';
export * from './core/log';
export * from './core/step';
export * from './core/stack-log.builder';
export * from './core/bind-log.builder';
export * from './utils/color.util';
export * from './utils/write-default-local.util';
export * from './utils/ensure-local-folder.util';
export * from './utils/get-folder-name.util';
export * from './utils/get-file-name.util';
export * from './utils/get-file-name.util';
export * from './utils/step-from-axios-error.util';
export * from './core/providers';
export * from './core/http-provider';
export * from './core/s3-provider';
export * from './core/auto-publish-log.builder';
export * from './utils/decrypt-keys.util';
export * from './utils/decrypt.util';
export * from './utils/encrypt-keys.util';
export * from './utils/encrypt-string.util';
export * from './utils/encrypt.util';
export * from './utils/encrypt-data.util';
export * from './utils/delete-expired-file.util';
export * from './utils/get-logs-dirname.util';
export * from './utils/mask-sub-key.util';
export * from './utils/mask-attribute.util';
export * from './utils/mask-step-data.util';
