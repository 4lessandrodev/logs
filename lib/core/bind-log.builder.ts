import { Log } from "./log";
import { BinderMiddleware, NextFunctions, Requests, Responses } from "../types";
import getLogDataFromRequest from "../utils/get-log-data-from-request.util";

/**
 * @description Bind log instance to request.
 * @returns execute next()
 * @summary you can access log from request like example:
 * @example req.log
 */
export const bindLog = (): BinderMiddleware => {
    return (req: Requests, _: Responses, next: NextFunctions): void => {
        const { name, ip, origin, uid } = getLogDataFromRequest(req);
        req.log = Log.init({ name, ip, origin, uid })
        next();
    }
}

export default bindLog;
