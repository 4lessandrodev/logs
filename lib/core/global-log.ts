import { Log } from "./log";

/**
 * @description Global log instance
 */
export abstract class GlobalLog {
    private static instance: Log;

    /**
     * @description Get or Create a log singleton instance
     * @param props as options to define id and name for log object.
     * @returns instance of Log.
     */
    public static singleton(props?: { uid?: string, name?: string }): Log {
        if(!GlobalLog.instance){
            const name = props?.name ?? 'global-log';
            const uid = props?.uid;
            GlobalLog.instance = Log.init({ name, uid }) as Log;
        }
        return GlobalLog.instance;
    }
}

export default GlobalLog;
