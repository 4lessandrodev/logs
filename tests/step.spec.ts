import { SProps } from "../lib/types";
import { Log, Step } from "../lib/core";

describe('step', () => {
    it('should create a step with success', () => {
        const keys: Array<keyof SProps> = [
            "uid",
            "name",
            "tags",
            "url",
            "stack",
            "data",
            "statusCode",
            "message",
            "type",
            "method",
            "createdAt"
        ];
        const step = Step.create({});
        expect.assertions(keys.length + 1);
        expect(Object.keys(step)).toEqual(keys);
        let i = 0;
        while (keys[i]) {
            expect(step[keys[i]]).toBeDefined();
            i++;
        }
    });

    it('should execute addTag', () => {
        const step = Step.create({});
        const result = step.addTag('testing')
        expect(result.tags).toHaveLength(1);
    });
    it('should execute addTags', () => {
        const step = Step.create({});
        const result = step.addTags(['1', '2'])
        expect(result.tags).toHaveLength(2);
    });
    it('should execute setName', () => {
        const step = Step.create({});
        const result = step.setName("changed")
        expect(result.name).toBe('changed')
    });
    it('should execute setMethod', () => {
        const step = Step.create({});
        const result = step.setMethod('GET')
        expect(result.method).toBe('GET')
    });
    it('should execute setStack', () => {
        const step = Step.create({});
        const result = step.setStack('internal server error line 10');
        expect(result.stack).toBe('internal server error line 10')
    });
    it('should execute setMessage', () => {
        const step = Step.create({});
        const result = step.setMessage('Timeout')
        expect(result.message).toBe('Timeout')
    });
    it('should execute setStatusCode', () => {
        const step = Step.create({});
        const result = step.setStatusCode(401);
        expect(result.statusCode).toBe(401);
    });
    it('should execute setData', () => {
        const step = Step.create({});
        const result = step.setData('{ "email": "sample@mail.com" }');
        expect(result.data).toBe('{ "email": "sample@mail.com" }');
    });
    it('should execute setUid', () => {
        const step = Step.create({});
        const result = step.setUid('000001')
        expect(result.uid).toBe('000001');
    });
    it('should execute setURL', () => {
        const step = Step.create({});
        const result = step.setURL('https://test.com.br')
        expect(result.url).toBe('https://test.com.br')
    });
    it('should create a step from error', () => {
        class ErrorUseCase {
            create(): Error {
                return new Error("simple error execution");
            }
        }

        const instance = new ErrorUseCase();
        const err = instance.create();
        expect(err).toBeInstanceOf(Error);

        const step = Step.catch(err);
        expect(step).toBeDefined();
    });

    it('should generate full log from axios', () => {

        class Testing {
            execute() {
                const axiosErr = { "message": "Request failed with status code 404", "name": "AxiosError", "stack": "AxiosError: Request failed with status code 404\n    at settle (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:1896:12)\n    at BrotliDecompress.handleStreamEnd (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:2940:11)\n    at BrotliDecompress.emit (node:events:525:35)\n    at endReadableNT (node:internal/streams/readable:1359:12)\n    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)", "config": { "transitional": { "silentJSONParsing": true, "forcedJSONParsing": true, "clarifyTimeoutError": false }, "adapter": ["xhr", "http"], "transformRequest": [null], "transformResponse": [null], "timeout": 0, "xsrfCookieName": "XSRF-TOKEN", "xsrfHeaderName": "X-XSRF-TOKEN", "maxContentLength": -1, "maxBodyLength": -1, "env": {}, "headers": { "Accept": "application/json, text/plain, */*", "Content-Type": "application/json", "User-Agent": "axios/1.3.2", "Content-Length": "47", "Accept-Encoding": "gzip, compress, deflate, br", "uid": "508791f9-9063-463d-a54b-cb6a777870af" }, "method": "post", "url": "https://postback-4dev.onrender.com/inv/not-found", "data": "{\"email\":\"test@domain.com\",\"password\":\"123456\"}" }, "code": "ERR_BAD_REQUEST", "status": 404 };
                const result = Step.catch(axiosErr);
                return result;
            }
        }

        const instance = new Testing();
        const result = instance.execute();

        expect(result).toEqual({
            "createdAt": expect.any(Date),
            "data": `{
  \"requestData\": {
    \"email\": \"test@domain.com\",
    \"password\": \"123456\"
  },
  \"responseData\": {}
}`,
            "message": "Request failed with status code 404",
            "method": "POST",
            "name": "Object",
            "stack": `AxiosError: Request failed with status code 404
    at settle (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:1896:12)
    at BrotliDecompress.handleStreamEnd (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:2940:11)
    at BrotliDecompress.emit (node:events:525:35)
    at endReadableNT (node:internal/streams/readable:1359:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)`,
            "statusCode": 500,
            "tags": [
                "email",
                "password",
            ],
            "type": "fatal",
            "uid": "508791f9-9063-463d-a54b-cb6a777870af",
            "url": "https://postback-4dev.onrender.com/inv/not-found"
        });

        expect(result).toMatchSnapshot({ createdAt: expect.any(Date) });

    });

    it('should generate full log from axios and remove password', () => {

        class Testing {
            execute() {
                const axiosErr = { "message": "Request failed with status code 404", "name": "AxiosError", "stack": "AxiosError: Request failed with status code 404\n    at settle (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:1896:12)\n    at BrotliDecompress.handleStreamEnd (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:2940:11)\n    at BrotliDecompress.emit (node:events:525:35)\n    at endReadableNT (node:internal/streams/readable:1359:12)\n    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)", "config": { "transitional": { "silentJSONParsing": true, "forcedJSONParsing": true, "clarifyTimeoutError": false }, "adapter": ["xhr", "http"], "transformRequest": [null], "transformResponse": [null], "timeout": 0, "xsrfCookieName": "XSRF-TOKEN", "xsrfHeaderName": "X-XSRF-TOKEN", "maxContentLength": -1, "maxBodyLength": -1, "env": {}, "headers": { "Accept": "application/json, text/plain, */*", "Content-Type": "application/json", "User-Agent": "axios/1.3.2", "Content-Length": "47", "Accept-Encoding": "gzip, compress, deflate, br", "uid": "508791f9-9063-463d-a54b-cb6a777870af" }, "method": "post", "url": "https://postback-4dev.onrender.com/inv/not-found", "data": "{\"email\":\"test@domain.com\",\"password\":\"123456\"}" }, "code": "ERR_BAD_REQUEST", "status": 404 };
                const result = Step.catch(axiosErr, { remove: ['password'] });
                return result;
            }
        }

        const instance = new Testing();
        const result = instance.execute();

        expect(result).toEqual({
            "createdAt": expect.any(Date),
            "data": `{
  \"requestData\": {
    \"email\": \"test@domain.com\"
  },
  \"responseData\": {}
}`,
            "message": "Request failed with status code 404",
            "method": "POST",
            "name": "Object",
            "stack": `AxiosError: Request failed with status code 404
    at settle (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:1896:12)
    at BrotliDecompress.handleStreamEnd (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:2940:11)
    at BrotliDecompress.emit (node:events:525:35)
    at endReadableNT (node:internal/streams/readable:1359:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)`,
            "statusCode": 500,
            "tags": [
                "email"
            ],
            "type": "fatal",
            "uid": "508791f9-9063-463d-a54b-cb6a777870af",
            "url": "https://postback-4dev.onrender.com/inv/not-found"
        });
        expect(result).toMatchSnapshot({ createdAt: expect.any(Date) });
    })
});
