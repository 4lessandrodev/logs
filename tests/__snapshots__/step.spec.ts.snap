// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`step should generate full log from axios 1`] = `
{
  "createdAt": Any<Date>,
  "data": "{
  "requestData": {
    "email": "test@domain.com",
    "password": "123456"
  },
  "responseData": {}
}",
  "message": "Request failed with status code 404",
  "method": "POST",
  "name": "Testing.execute",
  "stack": "AxiosError: Request failed with status code 404
    at settle (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:1896:12)
    at BrotliDecompress.handleStreamEnd (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:2940:11)
    at BrotliDecompress.emit (node:events:525:35)
    at endReadableNT (node:internal/streams/readable:1359:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)",
  "statusCode": 500,
  "tags": [
    "email",
    "password",
  ],
  "type": "fatal",
  "uid": "508791f9-9063-463d-a54b-cb6a777870af",
  "url": "https://postback-4dev.onrender.com/inv/not-found",
}
`;

exports[`step should generate full log from axios and remove password 1`] = `
{
  "createdAt": Any<Date>,
  "data": "{
  "requestData": {
    "email": "test@domain.com"
  },
  "responseData": {}
}",
  "message": "Request failed with status code 404",
  "method": "POST",
  "name": "Testing.execute",
  "stack": "AxiosError: Request failed with status code 404
    at settle (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:1896:12)
    at BrotliDecompress.handleStreamEnd (/home/alessandro/Workspace/logs-app/node_modules/axios/dist/node/axios.cjs:2940:11)
    at BrotliDecompress.emit (node:events:525:35)
    at endReadableNT (node:internal/streams/readable:1359:12)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)",
  "statusCode": 500,
  "tags": [
    "email",
  ],
  "type": "fatal",
  "uid": "508791f9-9063-463d-a54b-cb6a777870af",
  "url": "https://postback-4dev.onrender.com/inv/not-found",
}
`;
